### PR TITLE
Added initial eqsat front end pass

### PR DESCRIPTION
**Pull request by jianyicheng**
This PR addresses https://github.com/xdslproject/xdsl/issues/3170:
* Added initial front end pass `eqsat-create-eclasses` for the minimal example
* Updated the test case to match the pass

### CODE REVIEW

File: xdsl/tools/command_line_tool.py [hunk 1]
```diff
@@ -405,6 +405,11 @@ def get_stencil_bufferize():
 
         return stencil_bufferize.StencilBufferize
 
+    def get_create_eclasses():
```
Comment (superlopuh) [inline xdsl/tools/command_line_tool.py:408]:
```suggestion
    def get_eqsat_create_eclasses():
```

File: xdsl/transforms/eqsat_create_eclasses.py [hunk 2]
```diff
@@ -1,32 +1,91 @@
...
 
 def insert_eclass_ops(block: Block):
+    # Insert eqsat.eclass for each operation
     for op in block.ops:
         results = op.results
         if len(results) != 1:
-            raise DiagnosticException("Ops with non-single results not handled")
+            continue
+            # TODO: ignore operations with no result for now
+            # raise DiagnosticException("Ops with non-single results not handled")
```
Comment (superlopuh) [inline xdsl/transforms/eqsat_create_eclasses.py:22]: for our WIP branch, this is fine, but for main, we generally avoid TODO comments without associated tasks to track them, in a comment in the code. In this case, I'd just raise the diagnostic exception, and add a filecheck test to track that the diagnostic error is raised as expected (with a `--verify-diagnostics` flag passed to `xdsl-opt`). With that setup we tend not to file issues, and just fix the test when we come across the error.

File: xdsl/transforms/eqsat_create_eclasses.py [hunk 3]
```diff
@@ -1,32 +1,91 @@
...
 
         eclass_op = eqsat.EClassOp(results[0])
-        Rewriter.insert_op(op, InsertPoint.after(eclass_op))
-        eclass_op.results[0].replace_by(eclass_op.result)
+        insertion_point = InsertPoint.after(op)
+        Rewriter.insert_op(eclass_op, insertion_point)
+        results[0].replace_by(eclass_op.results[0])
+        # Redirect eclassop operand back to the original value
+        # TODO: do we need a `replace_by_except` function, e.g.
+        # eclass_op.result.replace_by_except(eclass_op.results[0], [eclass_op.result])
```
Comment (superlopuh) [inline xdsl/transforms/eqsat_create_eclasses.py:30]: I'm not sure what MLIR has for this, but this works ok, no?

Comment (jianyicheng) [reply xdsl/transforms/eqsat_create_eclasses.py:30]: I am thinking of this: https://mlir.llvm.org/doxygen/classmlir_1_1Operation.html#a6f5567fd62feeaf0cbba4877732c512a

File: xdsl/transforms/eqsat_create_eclasses.py (hunk 2 above)
Comment (jianyicheng) [reply xdsl/transforms/eqsat_create_eclasses.py:22]: Fixed here: https://github.com/xdslproject/xdsl/pull/3189

**Pull request merged**